js/ui/iconGrid.js

```diff
@@ -786,13 +786,36 @@ const IconGrid = new Lang.Class({
             let sourceActor = children[sourceIndex];
             let actorOffset;
 
-            if (targetIndex >= children.length) {
-                // calculate the position of the new slot
-                let oldBox = sourceActor.allocation;
-                let newBox = this._calculateChildBox(sourceActor, sourceActor.x, sourceActor.y, box);
-                actorOffset = [newBox.x1 - oldBox.x1, newBox.y1 - oldBox.y1];
-            } else {
+            // There are two relevant cases:
+            //  1. We want to move icons between well-known positions. This usually
+            //     happens when the user moves an icon to inside a folder, and we
+            //     have to move the icons back.
+            //  2. We don't know the position. This happens when moving from inside
+            //     a folder to the main grid. The icons will be pushed forward by
+            //     one position, and the last icon will go to the unknown position.
+            if (targetIndex < children.length) {
```

Comment:
Not a big deal, but the diff would be less confusing if you kept the conditions in the same order than the original code (if branch -> Case 2 / else branch -> Case 1), and maybe reorder the comment above